core/types/transaction.go

```diff
@@ -135,58 +135,69 @@ func DecodeTransaction(s *rlp.Stream) (Transaction, error) {
 		}
 		return tx, nil
 	}
-	if rlp.String == kind {
-		s.NewList(size) // Hack - convert String (envelope) into List
+	if rlp.String != kind {
+		return nil, fmt.Errorf("Not an RLP encoded transaction. If this is a canonical encoded transaction, use UnmarshalTransactionFromBinary instead. Got %v for kind, expected String", kind)
 	}
+	// Decode the EIP-2718 typed TX envelope.
 	var b []byte
 	if b, err = s.Bytes(); err != nil {
 		return nil, err
 	}
-	if len(b) != 1 {
-		return nil, fmt.Errorf("%w, got %d bytes", rlp.ErrWrongTxTypePrefix, len(b))
+	if len(b) == 0 {
+		return nil, rlp.EOL
 	}
-	var tx Transaction
-	switch b[0] {
+	return UnmarshalTransactionFromBinary(b)
+}
+
+// DecodeTransaction decodes a transaction either in RLP or canonical format
+func DecodeTransaction(data []byte) (Transaction, error) {
+	if len(data) == 0 {
+		return nil, io.EOF
+	}
+	if data[0] < 0x80 {
```

Comment:
curious what is the number or number range that represents RLP encoding? Wondering if there's unwanted consequences using `<` comparison

Reply:
see https://github.com/ledgerwatch/erigon/blob/5fd3536ef0a26b33e44a711884a4b96e3ce8181f/rlp/decode.go#L1002

RLP encoded transactions never start with a single byte, and are either of type List (for legacy Txs) or String (for newer txs which are stored in a "transaction envelope").  So if we see the single by there we know it's a canonical encoded tx which always starts with a single byte tx-type identifier.

Yes this is still a hack and this function shouldn't be overloaded at all IMO, but fixing all the callers and figuring out which type of data they are decoding turned out to be non-trivial so I left it at this.